### ISSUE TITLE
recognize_beam_batch throws an error AttributeError: 'list' object has no attribute 'view' because a_prev is None (fails for all coverage, location, multi_head variants)

### ISSUE THREAD

**Issue by sknadig** [espnet/espnet#665]
In [decoders.py#L457](https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/decoders.py#L457) `a_prev` is not updated, it is always `None`.
This results in the [condition for handling list](https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/decoders.py#L515) `if not isinstance(a_prev, list)` to be True always.
`att_w` will be a list for thest attention mechanisms, `att_w.view(n_bb, -1)` fails.
`AttributeError: 'list' object has no attribute 'view'`

One solution is to use `att_w` instead of `a_prev` while calling `self.att` at [decoders.py#L457](https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/decoders.py#L457)
`att_c, att_w = self.att[att_idx](exp_h, exp_hlens, self.dropout_dec[0](z_prev[0]), a_prev)`

I think this is related to #510

**Comment by sknadig** [comment]
For location_recurrent type attention, the return values are `return c, (w, (att_h, att_c))`
This needs to be handled separately.
```
if isinstance(att_w, tuple):
    #handle the case of location_recurrent when return is a tuple
    a_prev = torch.index_select(att_w[0].view(n_bb, -1), 0, vidx)
elif not isinstance(att_w, list):
    a_prev = torch.index_select(att_w.view(n_bb, -1), 0, vidx)
else:
    # adapt for multi-head attention
    a_prev = [torch.index_select(att_w_one.view(n_bb, -1), 0, vidx) for att_w_one in att_w]
```

**Comment by sw005320** [comment]
I see, thanks. @kan-bayashi, could you fix it? Probably better to add the test for this.

**Comment by kan-bayashi** [comment]
Fixed (#681)
